dartagnan/src/main/java/com/dat3m/dartagnan/program/analysis/alias/InclusionBasedPointerAnalysis.java

```diff
@@ -0,0 +1,897 @@
+    // A variable cannot be removed, if some event maps to it and there are multiple replacements.
+    // In this case, the mapping stays but all outgoing edges are removed from that variable.
+    private void replace(Variable old, Offset<Variable> replacement) {
+        assert old != replacement.base;
+        assert !objectVariables.containsValue(old);
+        totalReplacements++;
+        logger.trace("{} -> {}", old, replacement);
+        // likely / most frequent case
+        addressVariables.replaceAll((k, v) -> v.base != old ? v : join(replacement, v.offset, v.alignment));
+        registerVariables.replaceAll((k, v) -> v.base != old ? v : join(replacement, v.offset, v.alignment));
+        for (final Variable out : old.seeAlso) {
+            out.includes.replaceAll(e -> e.base != old ? e : join(replacement, e.offset, e.alignment));
+            out.loads.replaceAll(e -> e.base != old ? e : join(replacement, e.offset, e.alignment));
+            out.stores.replaceAll(e -> e.base.base != old ? e :
+                    new Offset<>(join(replacement, e.base.offset, e.base.alignment), e.offset, e.alignment));
+        }
+        replacement.base.seeAlso.addAll(old.seeAlso);
+        for (final Offset<Variable> edge : old.includes) {
+            edge.base.seeAlso.remove(old);
+            edge.base.seeAlso.add(replacement.base);
+        }
+        // Redirect load and store relationships.
+        // This could enable more communications, but replacement.
+        for (final Offset<Variable> load : old.loads) {
+            addInto(replacement.base.loads, join(load, replacement.offset, replacement.alignment));
+            load.base.seeAlso.add(replacement.base);
+        }
+        for (final Offset<Offset<Variable>> store : old.stores) {
+            addInto(replacement.base.stores, join(store, replacement.offset, replacement.alignment));
+            store.base.base.seeAlso.add(replacement.base);
+        }
+        old.seeAlso.clear();
+        old.includes.clear();
+        old.loads.clear();
+        old.stores.clear();
+    }
+
+    // Find "may read from" relationships and deduce new 'includes' edges for the internal graph.
+    private void processCommunication(Collection<Offset<Offset<Variable>>> stores, Collection<Offset<Variable>> loads) {
+        logger.trace("{} vs {}", stores, loads);
+        if (loads.isEmpty()) {
+            return;
+        }
+        for (final Offset<Offset<Variable>> store : stores) {
+            for (final Offset<Variable> load : loads) {
+                if (overlaps(load.offset - store.offset, store.alignment, load.alignment)) {
+                    addEdge(store.base.base, load.base, store.base.offset, store.base.alignment);
+                }
+            }
+        }
+    }
+
+    private record Offset<Base>(Base base, int offset, List<Integer> alignment) {
+        @Override public String toString() { return String.format("%s+%d+%sx", base, offset, alignment); }
+    }
```

Comment:
I would move this declaration directly before or after `Variable` because those two classes are very tightly related.